tests/virt_autotest/validate_system_health.pm

```diff
@@ -0,0 +1,73 @@
+# SUSE's openQA tests
+#
+# Copyright 2012-2022 SUSE LLC
+# SPDX-License-Identifier: FSFAP
+#
+# Summary: Do a basic examination to the host and guests after tests run
+# Maintainer: Julie CAO <JCao@suse.com>, qe-virt@suse.de
+
+use base "virt_feature_test_base";
+use strict;
+use warnings;
+use testapi;
+use utils;
+use ipmi_backend_utils qw(reconnect_when_ssh_console_broken);
+use Utils::Architectures;
+use virt_autotest::common;
+use virt_autotest::utils qw(start_guests check_host_health check_guest_health);
+use virt_utils qw(collect_host_and_guest_logs);
+
+sub prepare_run_test {
+    reconnect_when_ssh_console_broken unless defined(script_run("rm -f /root/{commands_history,commands_failure}", die_on_timeout => 0));
+    script_run("history -c");
+}
+
+sub run_test {
+
+    my $self = shift;
+    my %health_status = ();
+    @health_status{(keys %virt_autotest::common::guests)} = ();
+    $health_status{host} = check_host_health;
+    foreach my $guest (grep { $_ ne 'host' } keys %health_status) {
+        script_run("virsh start $guest", die_on_timeout => 0);
+        if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 2, retry => 60, die => 0) == 0) {
+            $health_status{$guest} = check_guest_health($guest);
+        }
+        else {
+            $health_status{$guest} = "Unreachable";
+        }
+        record_info("Failed guest", "$guest: $health_status{$guest}", result => 'fail') unless $health_status{$guest} eq 'pass';
+    }
+    my @unhealthy_list = grep { $health_status{$_} ne 'pass' } keys %health_status;
+
+    # Upload logs on x86_64
+    if (is_x86_64 && (get_var('FORCE_UPLOAD_LOGS') || @unhealthy_list != 0)) {
+        if (get_var('FORCE_UPLOAD_LOGS')) {
+            collect_host_and_guest_logs(join(' ', grep { $_ ne 'host' } keys %health_status));
+        }
+        else {
+            collect_host_and_guest_logs(join(' ', grep { $_ ne 'host' } @unhealthy_list));
+        }
+        $self->upload_coredumps;
+        save_screenshot;
+    }
+    die "Host or guests are not healthy!" unless @unhealthy_list == 0;
+}
+
+sub post_fail_hook {
```

Comment:
I think you just _SKIP_POST_FAIL_HOOKS to do everything in main body of the module. So there will  be no duplicates in any form. You can still mark the module as fail.

Reply:
Do not these logs need to be uploaded if collect_logs fails? /var/log/clean_up_virt_logs.log... 

I don't know much about these logs, so just leave them as they are. If you think it is not required, I will remove them.

Please help list which statements are useless:
```
    $self->junit_log_provision((caller(0))[3]);
    reconnect_when_ssh_console_broken unless defined(script_run("history -w /root/commands_history", die_on_timeout => 0));
    $self->upload_coredumps;    //I would like to keep it to upload coredump even collect_logs fails.
    upload_logs("/var/log/clean_up_virt_logs.log");
    upload_logs("/var/log/guest_console_monitor.log");
    script_run("rm -f -r /var/log/clean_up_virt_logs.log /var/log/guest_console_monitor.log");
```

Reply:
I mean you do not need to enter into post_fail_hook to avoid duplicate operations. But call still group some codes into a separate function, for example, upload_and_cleanup, to be used and called anywhere. I will have a deep look later.

Reply:
@Julie-CAO  @waynechen55  How about your agreement for this open conversation?

Reply:
The PR has lasted for too much time. Subsequent tests and more PRs are blocked. Any sugguestion or improvement, if not urgent, please create a ticket. I'll consider in the future. Please merge the PR ASAP. My another fix to SRIOV tests is in progress which is being blockd by this PR.

Reply:
I have no more comments. I think using post_fail_hook is acceptable.